python-package/lightgbm/sklearn.py

```diff
@@ -862,12 +906,29 @@ def fit(
         params["metric"] = [metric for metric in params["metric"] if metric is not None]
 
         if not isinstance(X, (pd_DataFrame, dt_DataTable)):
-            _X, _y = _LGBMCheckXY(X, y, accept_sparse=True, force_all_finite=False, ensure_min_samples=2)
+            _X, _y = _LGBMValidateData(
+                self,
+                X,
+                y,
+                # Prevent scikit-learn from deleting or modifying attributes like 'feature_names_in_' and 'n_features_in_'.
+                # We prefer to expose these via @property, to be able to raise a NotFittedError if they're accessed on an
+                # unfitted model... and so don't want to take on the complexity of defining setters and deleters for those.
+                reset=True,
+                # allow any input type (this validation is done further down, in lgb.Dataset())
+                accept_sparse=True,
+                # do not raise an error if Inf of NaN values are found (LightGBM handles these internally)
+                ensure_all_finite=False,
+                # raise an error on 0-row and 1-row inputs
+                ensure_min_samples=2,
+            )
             if sample_weight is not None:
                 sample_weight = _LGBMCheckSampleWeight(sample_weight, _X)
         else:
             _X, _y = X, y
 
+            # for other data types, setting n_features_in_ is handled by _LGBMValidateData() in the branch above
+            self.n_features_in_ = _X.shape[1]
```

Comment:
```suggestion
            self._n_features_in = _X.shape[1]
```

Reply:
I intentionally used `n_features_in_` (via the new setter I added for it in this PR) to match what `scikit-learn` does

https://github.com/scikit-learn/scikit-learn/blob/35f106c51c07a3eafebfa88759ac8227369d2be1/sklearn/utils/validation.py#L2758

Was thinking that there's no functional difference, but that doing that would make it a little easier for contributors who are familiar with `scikit-learn` and writing `scikit-learn` estimators to understand what this is doing.

Hearing that, do you still think it should be `_n_features_in` instead? If so I'll happily change it, I don't feel strongly.

Reply:
> I don't feel strongly.

Me too.

OK, fine! I simply wanted to avoid some overhead for setter call.